driver-core/src/main/com/mongodb/internal/async/AsyncRunnable.java

```diff
@@ -0,0 +1,188 @@
+                    callback.onResult(v, e);
+                } catch (Throwable t) {
+                    throw new CallbackThrew("Unexpected Throwable thrown from callback: ", e);
+                }
+            });
+        } catch (CallbackThrew t) {
+            // ignore
+        } catch (Throwable t) {
+            callback.onResult(null, t);
+        }
+    }
+
+    /**
+     * Must be invoked at end of async chain
+     * @param runnable the sync code to invoke (under non-exceptional flow)
+     *                 prior to the callback
+     * @param callback the callback provided by the method the chain is used in
+     */
+    default void thenRunAndFinish(final Runnable runnable, final SingleResultCallback<Void> callback) {
+        this.finish((r, e) -> {
+            if (e != null) {
+                callback.onResult(null, e);
+                return;
+            }
+            try {
+                runnable.run();
+            } catch (Throwable t) {
+                callback.onResult(null, t);
+                return;
+            }
+            callback.onResult(null, null);
+        });
+    }
+
+    /**
+     * See {@link #thenRunAndFinish(Runnable, SingleResultCallback)}, but the runnable
+     * will always be executed, including on the exceptional path.
+     * @param runnable the runnable
+     * @param callback the callback
+     */
+    default void thenAlwaysRunAndFinish(final Runnable runnable, final SingleResultCallback<Void> callback) {
+        this.finish((r, e) -> {
+            try {
+                runnable.run();
+            } catch (Throwable t) {
+                callback.onResult(null, t);
+                return;
+            }
+            callback.onResult(r, e);
+        });
+    }
+
+    /**
+     * @param runnable The async runnable to run after this one
+     * @return the composition of this and the runnable
+     */
+    default AsyncRunnable thenRun(final AsyncRunnable runnable) {
+        return (c) -> {
+            this.finish((r, e) -> {
+                if (e != null) {
+                    c.onResult(null, e);
+                    return;
+                }
+                try {
+                    runnable.finish(c);
+                } catch (Throwable t) {
+                    c.onResult(null, t);
+                }
+            });
+        };
+    }
+
+    /**
+     * @param supplier The supplier to supply using after this runnable.
+     * @return the composition of this runnable and the supplier
+     * @param <T> The return type of the supplier
+     */
+    default <T> AsyncSupplier<T> thenSupply(final AsyncSupplier<T> supplier) {
+        return (c) -> {
+            this.finish((r, e) -> {
+                if (e != null) {
+                    c.onResult(null, e);
+                    return;
+                }
+                try {
+                    supplier.finish(c);
+                } catch (Throwable t) {
+                    c.onResult(null, t);
+                }
```

Comment:
No methods should write any protecting code when executing `AsyncSupplier` because `AsyncSupplier.finish` already has such code

```suggestion
                supplier.finish(c);
```